app/controllers/LongerCommodityCodeController.scala

```diff
@@ -53,79 +52,77 @@ class LongerCommodityCodeController @Inject() (
-
-      originalComcodeOpt match {
-        case Some(shortCommodity) if commodityCodeSansTrailingZeros(shortCommodity).length == 6 =>
-          val shortCode = commodityCodeSansTrailingZeros(shortCommodity)
+      val shortComcodeOpt         = getShortCommodityCodeOpt(recordId, request, validLength)
+      val currentlyCategorisedOpt =
+        request.userAnswers
+          .get(RecordCategorisationsQuery)
+          .flatMap(_.records.get(recordId))
+          .map(_.commodityCode.padTo(maxValidLength, "0").mkString)
+
+      shortComcodeOpt match {
+        case Some(shortComcode) if shortComcode.length == validLength =>
           form
             .bindFromRequest()
             .fold(
-              formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, shortCode, recordId))),
+              formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, shortComcode, recordId))),
               value => {
-                val longCommodityCode   = s"$shortCode$value"
-                val shouldRedirectToCya = latestComcodeOpt.exists(latest =>
-                  shortCode.concat(value) == latest &&
-                    categorisationInfoOpt.exists(_.latestDoesNotMatchOriginal)
+                val longCommodityCode   = (shortComcode + value).padTo(maxValidLength, "0").mkString
+                val shouldRedirectToCya = currentlyCategorisedOpt.contains(longCommodityCode) && mode == CheckMode
+                updateAnswersAndProceedWithJourney(
+                  mode,
+                  recordId,
+                  value,
+                  longCommodityCode,
+                  shortComcode,
+                  shouldRedirectToCya
                 )
-
-                if (shouldRedirectToCya) {
-                  Future.successful(
-                    Redirect(
-                      navigator
-                        .nextPage(LongerCommodityCodePage(recordId, shouldRedirectToCya), mode, request.userAnswers)
-                    )
-                  )
-                } else {
-                  saveAnswerAndProceedWithJourney(mode, recordId, value, longCommodityCode, shortCode)
-                }
               }
             )
-        case _                                                                                  => Future.successful(Redirect(routes.JourneyRecoveryController.onPageLoad().url))
+        case _                                                        =>
+          Future.successful(Redirect(routes.JourneyRecoveryController.onPageLoad().url))
       }
     }
 
-  private def saveAnswerAndProceedWithJourney(
+  private def getShortCommodityCodeOpt(
+    recordId: String,
+    request: DataRequest[AnyContent],
+    validLength: Int
+  ): Option[String] =
+    request.userAnswers
+      .get(RecordCategorisationsQuery)
+      .flatMap(_.records.get(recordId))
+      .flatMap(_.originalCommodityCode.map(_.reverse.dropWhile(_ == "0").reverse.padTo(validLength, "0").mkString))
```

Comment:
what about 6 digit commodity codes that end in a 0, wouldn't this make it 5 for example?

Reply:
thankfully the answer is no

`.padTo(validLength, "0")`

ensures it is 6